src/table-procedure/src/drop_database.rs

```diff
@@ -0,0 +1,334 @@
+// Copyright 2023 Greptime Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Procedure to drop a database.
+
+use async_trait::async_trait;
+use catalog::{CatalogManagerRef, DeregisterSchemaRequest};
+use common_procedure::error::SubprocedureFailedSnafu;
+use common_procedure::{
+    Context, Error, LockKey, Procedure, ProcedureId, ProcedureManager, ProcedureState,
+    ProcedureWithId, Result, Status,
+};
+use common_telemetry::logging;
+use serde::{Deserialize, Serialize};
+use snafu::{ensure, ResultExt};
+use table::engine::{EngineContext, TableEngineProcedureRef, TableReference};
+use table::requests::{DropDatabaseRequest, DropTableRequest};
+
+use crate::error::{
+    AccessCatalogSnafu, DeserializeProcedureSnafu, SchemaNotFoundSnafu, SerializeProcedureSnafu,
+};
+
+/// Procedure to drop a database.
+pub struct DropDatabaseProcedure {
+    data: DropDatabaseData,
+    catalog_manager: CatalogManagerRef,
+    engine_procedure: TableEngineProcedureRef,
+}
+
+#[async_trait]
+impl Procedure for DropDatabaseProcedure {
+    fn type_name(&self) -> &str {
+        Self::TYPE_NAME
+    }
+
+    async fn execute(&mut self, ctx: &Context) -> Result<Status> {
+        match self.data.state {
+            DropDatabaseState::Prepare => self.on_prepare().await,
+            DropDatabaseState::EngineDropAllTables => self.on_engine_drop_all_tables(ctx).await,
+            DropDatabaseState::RemoveFromCatalog => self.on_remove_from_catalog().await,
+        }
```

Comment:
@evenyag 

When we need to check if new tables are added to the schema?

Reply:
Even though we can check if new tables are added to the schema in `DropDatabaseProcedure`, I think we can't restore all talbes deleted from the schema.

Reply:
Sorry for the late reply.

> When do we need to check if new tables are added to the schema?

You could do this before `RemoveFromCatalog`, checking whether there are tables under the schema.

> I think we can't restore all talbes deleted from the schema.

Yes, we don't need to restore them. We return an error to users and also abort the procedure. if they still want to drop the schema, just rerun `drop database xxx;` again.

Reply:
Thanks for the reply.

Right now `DropDatabaseProcedure` and `CreateTableProcedure` doesn't lock the same resource.

So, I think this isn't thread-safe. 

For example, let's assume that `DropDatabaseProcedure` completes `DropDatabaseState::RemoveFromCatalog` before `CreateTableProcedure` completes `CreateTableState::RegisterCatalog`.

In this case, `CreateTableProcedure` should fail in the next step. `TableEngine` would still have the table, while `CatalogManager` wouldn't have reference to the table. So, No one can remove the table even if users execute `DropDropdatabaseProcedure` again.

https://github.com/GreptimeTeam/greptimedb/blob/4fd37d9d4e29d53c0930796f2e2be61a53f78775/src/table-procedure/src/create.rs#L49-L52

Reply:
> So, I think this isn't thread-safe.

Good catch. This is a potential issue. We could convert this into an issue and finally fix it after we refactor the lock manager and executor of the procedure framework.

> TableEngine would still have the table, while CatalogManager wouldn't have reference to the table and the schema.

In this case, we might not create the table successfully since the schema does not exist. But we still need a way to lock the schema in a shared or exclusive manner.

Reply:
>we might not create the table successfully since the schema does not exist.

I guess that we can replicate by the following steps:

1. `CreateTableState::EngineCreateTable`
2. `DropDatabaseState::EngineDropAllTables` (We can't remove the new table since `DropDatabaseState` doesn't have the new table name.)
3. `DropDatabaseState::RemoveFromCatalog` (But `TableEngine` still has the new table name.)
4. `CreateTableState::RegisterCatalog` (This step should fail since the schema doesn't exist.)

Reply:
>We could convert this into an issue and finally fix it after we refactor the lock manager and executor of the procedure framework.

Sounds good! I think we need to block this PR until we refactor them.